dedicated class and library for parameter handling

christianrauch opened the pull request:
The node declares and initialises all camera controls with their default control value.

However, some libcamera IPAs have a peculiar behaviour:
1. there is no guarantee that the `ControlInfo` with the min/max/def `ControlValue` uses the same data types as the `Control` they are assigned to, and
2. the default `ControlValue`s of multiple `Control` can be conflicting.

Example:
- `FrameDurationLimits` is of type `Control<Span<const int64_t, 2>>` (array with 2 elements) but its [bounds and default value are scalars](https://github.com/raspberrypi/libcamera/blob/v0.4.0%2Brpt20250213/src/ipa/rpi/common/ipa_base.cpp#L77-L78).
- For UVC devices `AeEnable` is on and `ExposureTime` is set by default. Setting both default values on their controls causes 
  ```
  ERROR V4L2 v4l2_device.cpp:390 /dev/video2[41:cap]: Unable to set controls: Permission denied
  ERROR UVC uvcvideo.cpp:385 Failed to set controls: -22
  ```

Previously, these kinds of cases were handled in functions `resolve_conflicts` and `check_conflicts`. This PR moves this functionality and the general parameter handling to the classes `ParameterHandler`, for general handling of camera controls and their respective ROS parameters, and `ParameterConflictHandler`, for handling and resolving conflicts of the camera controls.

christianrauch commented:
@samyarsadat Since you recently send PRs, this may be interesting for you. Would you be able to take a look and test this with your setup?

samyarsadat commented:
Of course, I would be happy to. I'll test it today.

samyarsadat commented:
I just did some testing, and everything seems to be working as expected. I did however notice one potential issue. 

There is code in `ParameterConflictHandler.cpp` to prevent `ExposureTime` from being set whilst `AeEnable` is enabled. This works as expected unless someone uses `1` instead of `true` when setting `AeEnable`. If set to `1`, ROS treats the `AeEnable` parameter as an integer. This is fine as far as libcamera is concerned, as it just treats int `1` as true, but `ParameterConflictHandler` only checks for conflicts when `AeEnable` is a Boolean (as far as I can understand). I set `AeEnable` to `1`, and then managed to set `ExpsoreTime` to `9999` without the parameter conflict warning appearing. Auto-exposure was still active. 

Here are the logs (with `--log-level debug`), in case it might be useful:
```
...
[DEBUG] [1739962245.831622025] [camera]: declare 'AwbEnable' (type: bool)
[DEBUG] [1739962245.831715635] [camera]: declare 'ColourGains' (type: double_array)
[WARN] [1739962245.831819115] [camera]: unknown control: SyncFrames (20014)
[DEBUG] [1739962245.831854411] [camera]: declare 'AwbMode' (type: integer)
[DEBUG] [1739962245.831922762] [camera]: declare 'HdrMode' (type: integer)
[DEBUG] [1739962245.831986372] [camera]: declare 'AeMeteringMode' (type: integer)
[DEBUG] [1739962245.832054205] [camera]: declare 'AfTrigger' (type: integer)
[DEBUG] [1739962245.832128111] [camera]: declare 'LensPosition' (type: double)
[DEBUG] [1739962245.832201184] [camera]: declare 'Contrast' (type: double)
[WARN] [1739962245.832278591] [camera]: unsupported control: AfWindows (35)
[DEBUG] [1739962245.832313053] [camera]: declare 'AeFlickerPeriod' (type: integer)
[DEBUG] [1739962245.832385163] [camera]: declare 'AeEnable' (type: bool)
[DEBUG] [1739962245.832468181] [camera]: declare 'AnalogueGain' (type: double)
[DEBUG] [1739962245.832550310] [camera]: declare 'AfPause' (type: integer)
[DEBUG] [1739962245.832629346] [camera]: declare 'AfMetering' (type: integer)
[DEBUG] [1739962245.832697900] [camera]: declare 'AfSpeed' (type: integer)
[DEBUG] [1739962245.832771048] [camera]: declare 'AfRange' (type: integer)
[DEBUG] [1739962245.832836713] [camera]: declare 'FrameDurationLimits' (type: integer_array)
[WARN] [1739962245.832911750] [camera]: unknown control: CnnEnableInputTensor (20007)
[DEBUG] [1739962245.832944249] [camera]: declare 'AfMode' (type: integer)
[DEBUG] [1739962245.833011378] [camera]: declare 'ExposureValue' (type: double)
[DEBUG] [1739962245.833073544] [camera]: declare 'AeFlickerMode' (type: integer)
[DEBUG] [1739962245.833138821] [camera]: declare 'ExposureTime' (type: integer)
[DEBUG] [1739962245.833203913] [camera]: declare 'AeExposureMode' (type: integer)
[WARN] [1739962245.833273097] [camera]: unknown control: SyncMode (20011)
[DEBUG] [1739962245.833299115] [camera]: declare 'Brightness' (type: double)
[DEBUG] [1739962245.833372244] [camera]: declare 'Sharpness' (type: double)
[WARN] [1739962245.833455076] [camera]: unknown control: NoiseReductionMode (10002)
[DEBUG] [1739962245.833482909] [camera]: declare 'StatsOutputEnable' (type: bool)
[DEBUG] [1739962245.833546834] [camera]: declare 'AeConstraintMode' (type: integer)
[DEBUG] [1739962245.833625852] [camera]: declare 'ScalerCrop' (type: integer_array)
[DEBUG] [1739962245.833694684] [camera]: declare 'Saturation' (type: double)
[DEBUG] [1739962245.833766628] [camera]: declare 'ColourTemperature' (type: integer)
[DEBUG] [1739962245.833938274] [camera]: setting double parameter Saturation to 1.000000
[DEBUG] [1739962245.833968126] [camera]: setting integer parameter AeConstraintMode to 0
[DEBUG] [1739962245.833987829] [camera]: setting double parameter Brightness to 0.000000
[DEBUG] [1739962245.834005736] [camera]: setting integer parameter AeExposureMode to 0
[DEBUG] [1739962245.834023329] [camera]: setting bool parameter StatsOutputEnable to false
[DEBUG] [1739962245.834042051] [camera]: setting integer parameter AeFlickerMode to 0
[DEBUG] [1739962245.834065495] [camera]: setting integer_array parameter ScalerCrop to [0, 0, 4608, 2592]
[DEBUG] [1739962245.834083402] [camera]: setting integer parameter AfRange to 0
[DEBUG] [1739962245.834100568] [camera]: setting integer parameter AfSpeed to 0
[DEBUG] [1739962245.834119272] [camera]: setting integer parameter AwbMode to 0
[DEBUG] [1739962245.834137883] [camera]: setting double parameter ExposureValue to 0.000000
[DEBUG] [1739962245.834156345] [camera]: setting integer parameter HdrMode to 0
[DEBUG] [1739962245.834174549] [camera]: setting bool parameter AeEnable to true
[DEBUG] [1739962245.834191882] [camera]: setting integer parameter AeMeteringMode to 0
[DEBUG] [1739962245.834208882] [camera]: setting integer parameter AfTrigger to 0
[DEBUG] [1739962245.834226733] [camera]: setting integer parameter AfPause to 0
[DEBUG] [1739962245.834278955] [camera]: setting double parameter LensPosition to 1.000000
[DEBUG] [1739962245.834297307] [camera]: setting double parameter Contrast to 1.000000
[DEBUG] [1739962245.834314714] [camera]: setting double parameter Sharpness to 1.000000
[DEBUG] [1739962245.834330066] [camera]: setting integer parameter AfMode to 0
[DEBUG] [1739962245.834346121] [camera]: setting integer parameter AfMetering to 0
[151:02:31.200121049] [41872]  WARN IPARPI ipa_base.cpp:1198 Could not set AF_PAUSE - no AF algorithm or not Continuous
[151:02:31.200153567] [41872]  WARN IPARPI ipa_base.cpp:1215 Could not set AF_TRIGGER - no AF algorithm or not Auto
...
[DEBUG] [1739962246.419236510] [camera]: applied control 'AfMode': 0
[DEBUG] [1739962246.419471248] [camera]: applied control 'Sharpness': 1.000000
[DEBUG] [1739962246.419525692] [camera]: applied control 'Contrast': 1.000000
[DEBUG] [1739962246.419554413] [camera]: applied control 'AfPause': 0
[DEBUG] [1739962246.419589598] [camera]: applied control 'AfTrigger': 0
[DEBUG] [1739962246.419630968] [camera]: applied control 'Saturation': 1.000000
[DEBUG] [1739962246.419664542] [camera]: applied control 'StatsOutputEnable': false
[DEBUG] [1739962246.419691282] [camera]: applied control 'AfSpeed': 0
[DEBUG] [1739962246.419718004] [camera]: applied control 'AeConstraintMode': 0
[DEBUG] [1739962246.419756004] [camera]: applied control 'ScalerCrop': (0, 0)/4608x2592
[DEBUG] [1739962246.419785448] [camera]: applied control 'AfMetering': 0
[DEBUG] [1739962246.419813614] [camera]: applied control 'AeExposureMode': 0
[DEBUG] [1739962246.419851706] [camera]: applied control 'LensPosition': 1.000000
[DEBUG] [1739962246.419879835] [camera]: applied control 'AeFlickerMode': 0
[DEBUG] [1739962246.419901983] [camera]: applied control 'AeMeteringMode': 0
[DEBUG] [1739962246.419926076] [camera]: applied control 'AfRange': 0
[DEBUG] [1739962246.419946483] [camera]: applied control 'ExposureValue': 0.000000
[DEBUG] [1739962246.419972834] [camera]: applied control 'AwbMode': 0
[DEBUG] [1739962246.419997093] [camera]: applied control 'Brightness': 0.000000
[DEBUG] [1739962246.420017278] [camera]: applied control 'HdrMode': 0
[DEBUG] [1739962246.420040222] [camera]: applied control 'AeEnable': true
[151:02:31.926537037] [41872]  WARN IPARPI ipa_base.cpp:1198 Could not set AF_PAUSE - no AF algorithm or not Continuous
[151:02:31.926564019] [41872]  WARN IPARPI ipa_base.cpp:1215 Could not set AF_TRIGGER - no AF algorithm or not Auto
(At this point, I ran "ros2 param set /camera AeEnable 1" from a second terminal)
[DEBUG] [1739962259.070236936] [rcl]: Subscription taking message
[DEBUG] [1739962259.070299639] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962259.073484914] [rcl]: Subscription taking message
[DEBUG] [1739962259.073545821] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962259.603960237] [rcl]: Service server taking service request
[DEBUG] [1739962259.604073939] [rcl]: Service take request succeeded: true
[DEBUG] [1739962259.604351991] [camera]: setting integer parameter AeEnable to 1
[DEBUG] [1739962259.604672117] [rcl]: Sending service response
[DEBUG] [1739962259.604919280] [rcl]: Subscription taking message
[DEBUG] [1739962259.604969446] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962259.638929975] [camera]: applied control 'AeEnable': true
(And here I ran "ros2 param set /camera ExposureTime 9999")
[DEBUG] [1739962270.000890725] [rcl]: Subscription taking message
[DEBUG] [1739962270.000959187] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962270.001178203] [rcl]: Subscription taking message
[DEBUG] [1739962270.001207944] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962270.531677938] [rcl]: Service server taking service request
[DEBUG] [1739962270.531743086] [rcl]: Service take request succeeded: true
[DEBUG] [1739962270.531841584] [camera]: setting integer parameter ExposureTime to 9999
[DEBUG] [1739962270.532073378] [rcl]: Sending service response
[DEBUG] [1739962270.532178747] [rcl]: Subscription taking message
[DEBUG] [1739962270.532236987] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962270.583271196] [camera]: applied control 'ExposureTime': 9999
```

Other than this, everything else seems to function correctly.

christianrauch commented:
Which commands did you exactly run?

If I set `AeEnable:=1` via the overrides:
```sh
ros2 run camera_ros camera_node --ros-args -p camera:=1 -p AeEnable:=1 -p ExposureTime:=5000 
```
I get a type check error:
```
terminate called after throwing an instance of 'rclcpp::ParameterTypeException'
  what():  expected [bool] got [integer]
[ros2run]: Aborted
```

If I run the node with defaults:
```sh
ros2 run camera_ros camera_node
```
and set `AeEnable:=0`:
```sh
ros2 param set /camera AeEnable 0
```
then it behaves as if `AeEnable` was set via boolean "true":
```
$ ros2 param dump /camera
/camera:
  ros__parameters:
    AeEnable: 0
    ExposureTime: 15600
```

I think on the command line, every value is a "string" initially and then interpreted into other values.

christianrauch commented:
Now I get what you meant.
Starting the node via:
```sh
ros2 run camera_ros camera_node
```
and setting `AeEnable` to `1` when it is already true:
```sh
ros2 param set /camera AeEnable 1
```
causes the issue:
```
ERROR V4L2 v4l2_device.cpp:390 /dev/video2[39:cap]: Unable to set controls: Permission denied
ERROR UVC uvcvideo.cpp:385 Failed to set controls: -22
```

samyarsadat commented:
> causes the issue:
> 
> ```
> ERROR V4L2 v4l2_device.cpp:390 /dev/video2[39:cap]: Unable to set controls: Permission denied
> ERROR UVC uvcvideo.cpp:385 Failed to set controls: -22
> ```

Well that's strange. I presume the behavior is different because I'm using a Raspberry Pi camera. 
With my setup, no errors occur when doing this. 

Here's a step-by-step explanation of what I did:
I first ran `ros2 run camera_ros camera_node --ros-args --log-level debug -p width:=2304 -p height:=1296 -p format:=RGB888 -p role:=video -p camera:=0` to start the node. I then ran `ros2 param set /camera AeEnable 1` from a second terminal, followed by `ros2 param set /camera ExposureTime 9999` (both returned success messages).

I've trimmed the log output of the node from before even further to only show the important parts:
```
(At this point, I ran "ros2 param set /camera AeEnable 1" from the other terminal)
[DEBUG] [1739962259.070236936] [rcl]: Subscription taking message
[DEBUG] [1739962259.070299639] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962259.073484914] [rcl]: Subscription taking message
[DEBUG] [1739962259.073545821] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962259.603960237] [rcl]: Service server taking service request
[DEBUG] [1739962259.604073939] [rcl]: Service take request succeeded: true
[DEBUG] [1739962259.604351991] [camera]: setting integer parameter AeEnable to 1
[DEBUG] [1739962259.604672117] [rcl]: Sending service response
[DEBUG] [1739962259.604919280] [rcl]: Subscription taking message
[DEBUG] [1739962259.604969446] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962259.638929975] [camera]: applied control 'AeEnable': true
(And here I ran "ros2 param set /camera ExposureTime 9999")
[DEBUG] [1739962270.000890725] [rcl]: Subscription taking message
[DEBUG] [1739962270.000959187] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962270.001178203] [rcl]: Subscription taking message
[DEBUG] [1739962270.001207944] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962270.531677938] [rcl]: Service server taking service request
[DEBUG] [1739962270.531743086] [rcl]: Service take request succeeded: true
[DEBUG] [1739962270.531841584] [camera]: setting integer parameter ExposureTime to 9999
[DEBUG] [1739962270.532073378] [rcl]: Sending service response
[DEBUG] [1739962270.532178747] [rcl]: Subscription taking message
[DEBUG] [1739962270.532236987] [rcl]: Subscription take succeeded: true
[DEBUG] [1739962270.583271196] [camera]: applied control 'ExposureTime': 9999
```

As you can see, int `1` is treated as `true`, even though the parameter type is integer. This makes it so that the `ExposureTime`/`AeEnable` conflict check is bypassed and allows me to set `ExposureTime` despite `AeEnable` effectively being set to `true`.

I presumed this behavior is unwanted, so I thought I'd point it out, but it doesn't cause any errors with my setup. `ExposureTime` seems to be ignored by the camera when `AeEnable` is set to `true` anyway.

christianrauch commented:
The problem was in `pv_to_cv`, where the function would convert the integer parameter to a bool control, because this cast is possible and we explicitly requested this, assuming the provided parameter type matches the declared parameter type. I moved the type check before the conversion and check against the original camera control type.

> I presumed this behavior is unwanted, so that's why I thought I'd point it out, but it doesn't cause any errors with my setup. `ExposureTime` seems to be ignored by the camera when `AeEnable` is set to `true` anyway.

This conversion behaviour should not be allowed since it defies the conflict checks. It may work for some cameras as their IPA is silently adjusting the configuration automatically. But for example the UVC will cancel the requests and this is then only visible from libcamera error messages on the terminal and thus hard to debug.

I added these manual type check because I had to make the ROS parameters dynamically typed in order to "unset" them, e.g. to unset `ExposureTime` when `AeEnable` is set to on/true. With static types, those checks are already done by ROS.

Thanks for testing and pointing that out. I pushed a fixed version to the PR branch. This should now work as expected.

samyarsadat commented:
> I added these manual type check because I had to make the ROS parameters dynamically typed in order to "unset" them, e.g. to unset `ExposureTime` when `AeEnable` is set to on/true. With static types, those checks are already done by ROS.

I see. I was wondering why ROS even allowed me to set an integer value for a Boolean parameter. That explains it.

> Thanks for testing and pointing that out. I pushed a fixed version to the PR branch. This should now work as expected.

No problem, glad to be of help. I tested again with the new changes, and everything is working as expected indeed.
